src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java

```diff
@@ -130,25 +152,33 @@ public void actionPerformed(@NotNull AnActionEvent e) {
       return;
     }
 
-    String language = courseProjectViewModel.languageProperty.get();
+    String language = Objects.requireNonNull(courseProjectViewModel.languageProperty.get());
     if (!tryCreateCourseFile(project, selectedCourseUrl, language)) {
       return;
     }
 
-    startAutoInstallsWithRestart(course, !courseProjectViewModel.userOptsOutOfSettings(), project);
+    Awaitable installDone = startAutoInstallsWithRestart(course, project);
+    Objects.requireNonNull(installDone); // because of checkstyle stupidity
 
     if (!tryImportProjectSettings(project, course)) {
       return;
     }
 
-    if (!courseProjectViewModel.userOptsOutOfSettings()) {
-      tryImportIdeSettings(course);
+    if (!courseProjectViewModel.userOptsOutOfSettings() && !tryImportIdeSettings(project, course)) {
+      return;
     }
 
     if (createCourseFile) {
       // The course file not created in testing.
       PluginSettings.getInstance().createUpdatingMainViewModel(project);
     }
+
+    executor.execute(() -> {
```

Comment:
OK, I understand the part with `dialog` goes and `notification` comes, but what's all the fuzz with `awaitable` and **async**?

Reply:
more I read it more I get a feeling this PR does something extra :stuck_out_tongue_winking_eye:

Reply:
In a way... or it is more like that transition to notifications make a certain problem apparent, and that problem I tried to fix here.

So, previously do-you-wanna-restart-dialog was shown after the auto-install modules were (asynchronously) installed *regardless of* whether settings were successfully imported or not. So even if settings import failed, the user was requested to restart. In practice, it wasn't a big issue as error dialog made it clear if something went wrong. With notifications, it's very likely that user wouldn't pay attention to an error notification that is shown at the same time with do-you-wanna-restart-dialog. So user just clicks "yes, restart now" and will never know that there was an error with settings import as restarting also clears notifications.

(I'm not sure if it was even possible that user clicks "do restart" before the import was completed. I guess, it wasn't, though.)

So now, do-you-wanna-restart-dialog is only shown after the settings have successfully imported (otherwise, a notification is shown). This is implemented (not perhaps best way but I tried to keep footprint small) so that after everything has gone well, line 176 spawns a new thread which 1) waits for auto-installment being completed, 2) shows the do-you-wanna-restart-dialog if needed.

Reply:
> (I'm not sure if it was even possible that user clicks "do restart" before the import was completed. I guess, it wasn't, though.)

I think the action (at least when launched from the A+ menu) runs on the UI thread, so the auto-installs (which use a write action) cannot finish before the action. Otherwise we would have an interesting situation where the user restarts before the action is completed.

However, I'm not sure if there are ways to use the action where it doesn't run on the UI thread, so I think this addition is a good guarantee against that (and of course showing the restart dialog when the IDE settings import failed is a bit silly)

Reply:
Btw is the separate waiting thread in the end needed for this reason, i.e. that the action completes?

Reply:
So this was reverted but I comment here:

> I think the action (at least when launched from the A+ menu) runs on the UI thread, so the auto-installs (which use a write action) cannot finish before the action.

This was my reasoning too but as said, it's vague and this guarantee is not clearly visible in the code.

> Btw is the separate waiting thread in the end needed for this reason, i.e. to let the action complete?

We don't want to wait in EDT. Yes, this action makes EDT wait on other parts (was the freezing of UI even reported by some of our testers) and we should get rid of those completely, so we don't want to add more waiting.